Apps/TimelineApp/core/sensor/src/main/java/uk/ac/cam/cares/jps/sensor/source/handler/SoundLevelHandler.java

```diff
@@ -105,13 +106,14 @@ public void run() {
     private void processAudioStream() {
         short[] buffer = new short[BUFFER_SIZE / 2];
         while (true) {
-            AudioRecord localAudioRecord = audioRecord; // local reference to prevent NullPointerException
-            if (localAudioRecord == null || localAudioRecord.getRecordingState() != AudioRecord.RECORDSTATE_RECORDING) {
-                break;
+            synchronized (audioRecordLock) {
+                if (audioRecord == null || audioRecord.getRecordingState() != AudioRecord.RECORDSTATE_RECORDING) {
+                    break;
+                }
             }
 
             try {
-                int readResult = localAudioRecord.read(buffer, 0, buffer.length);
+                int readResult = audioRecord.read(buffer, 0, buffer.length);
```

Comment:
better to include this in the synchronized block, otherwise it also risks from null pointer exception when audioRecordLock is released but picked up by stop()